File: 26-Dynamic-Programming-Part-II/cpp/04-rod-cutting-problem.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int RodCutting(vector<int>& price, int n) {
        vector<int> dp(n + 1, 0);
        for (int len = 1; len <= n; len++) {
            for (int cut = 1; cut <= len; cut++) {
                dp[len] = max(dp[len], price[cut - 1] + dp[len - cut]);
            }
        }
        return dp[n];
    }
};

File: 26-Dynamic-Programming-Part-II/python/04_rod_cutting_problem.py
from typing import List

class Solution:
    def RodCutting(self, price: List[int], n: int) -> int:
        dp = [0] * (n + 1)
        for length in range(1, n + 1):
            for cut in range(1, length + 1):
                dp[length] = max(dp[length], price[cut - 1] + dp[length - cut])
        return dp[n]
